platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,977 @@
+	IMB_MGR *mb_mgr;
+} crypto_local_t;
+
+static __thread crypto_local_t local;
+
+static
+odp_crypto_generic_session_t *alloc_session(void)
+{
+	odp_crypto_generic_session_t *session = NULL;
+
+	odp_spinlock_lock(&global->lock);
+	session = global->free;
+	if (session) {
+		global->free = session->next;
+		session->next = NULL;
+	}
+	odp_spinlock_unlock(&global->lock);
+
+	if (!session)
+		return NULL;
+
+	session->idx = session - global->sessions;
+
+	return session;
+}
+
+static
+void free_session(odp_crypto_generic_session_t *session)
+{
+	odp_spinlock_lock(&global->lock);
+	session->next = global->free;
+	global->free = session;
+	odp_spinlock_unlock(&global->lock);
+}
+
+static
+odp_crypto_packet_result_t *get_op_result_from_packet(odp_packet_t pkt)
+{
+	odp_packet_hdr_t *hdr = packet_hdr(pkt);
+
+	return &hdr->crypto_op_result;
+}
+
+static inline void set_crypto_op_result(odp_packet_t pkt,
+					odp_crypto_alg_err_t cipher_err,
+					odp_crypto_alg_err_t auth_err)
+{
+	odp_crypto_packet_result_t *op_result;
+
+	op_result = get_op_result_from_packet(pkt);
+	op_result->cipher_status.alg_err = cipher_err;
+	op_result->cipher_status.hw_err = ODP_CRYPTO_HW_ERR_NONE;
+	op_result->auth_status.alg_err = auth_err;
+	op_result->auth_status.hw_err = ODP_CRYPTO_HW_ERR_NONE;
+	op_result->ok = (cipher_err == ODP_CRYPTO_ALG_ERR_NONE &&
+			 auth_err   == ODP_CRYPTO_ALG_ERR_NONE);
+}
+
+static inline void set_crypto_op_result_ok(odp_packet_t pkt)
+{
+	set_crypto_op_result(pkt,
+			     ODP_CRYPTO_ALG_ERR_NONE,
+			     ODP_CRYPTO_ALG_ERR_NONE);
+}
+
+static void
+null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
+		    const odp_crypto_packet_op_param_t *param ODP_UNUSED,
+		    odp_crypto_generic_session_t *session ODP_UNUSED)
+{
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void zuc_eea3_cipher_op(odp_packet_t pkt,
+			const odp_crypto_packet_op_param_t *param,
+			odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr) {
+		iv_ptr = param->cipher_iv_ptr;
+	} else if (session->p.cipher_iv.data) {
+		iv_ptr = session->cipher.iv_data;
+	} else {
+		set_crypto_op_result(pkt,
+				     ODP_CRYPTO_ALG_ERR_DATA_SIZE,
+				     ODP_CRYPTO_ALG_ERR_NONE);
+		return;
+	}
+#else
+	iv_ptr = param->cipher_iv_ptr;
+	_ODP_ASSERT(session->p.cipher_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(seg_len < in_len)) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (session->p.cipher_key.length == 16) {
+		/* ZUC128 EEA3 */
+		IMB_ZUC_EEA3_1_BUFFER(mb_mgr, session->cipher.key_data,
```

Comment:
This might actually fail but this code does not notice it. I tested this PR a bit and noticed that with >8k cipher ranges the ciphering is actually not done but the operation silently fails. You should probably use imb_get_errno() to see if the operation actually succeeded.

Reply:
OK that's because check in function `zuc_eea3_1_buffer_aarch64` fails at this part
```        /* Check input data is in range of supported length */
        if (length < ZUC_MIN_BYTELEN ||
            length > ZUC_MAX_BYTELEN) {
                imb_set_errno(NULL, IMB_ERR_CIPH_LEN);
                return;
        }
```

I added this for all the return value checks, please review again.
 ```
	if (odp_unlikely(imb_get_errno(mb_mgr) != 0)) {
		set_crypto_op_result(pkt,
				     ODP_CRYPTO_ALG_ERR_DATA_SIZE,
				     ODP_CRYPTO_ALG_ERR_NONE);
		return;
	}
```

Reply:
The changed code still does not work. Now the error is checked and crypto op result is set, but the crypto op result may get overwritten by the second operation function. For each packet, you are calling both the auth and cipher function in some order and they cannot both set the op result. Now if either the cipher or auth function is the null function and the null function is called last, crypto op result will always be ok even if the actual crypto op function failed.

Reply:
Understood, added return values checks for both auth and cipher in function `crypto_int`.

Reply:
The code looks ok now, but the imb_get_errno() based error checking seems to be unreliable. In some cases I still get silent failures (imb_get_errno() not indiciating an error and this ODP not indicating an error) with too large packets. For instance, zuc256_eia3_1_buffer_aarch64_no_aesni() does not set IMB errno when it fails. It also does not clear the errno, so a previous error may be indicated after a (succesful, in reality) zuc256-eia operation.

I wonder if this is simply a bug in the library that will get fixed or whether ODP should have some hardcoded length check so that it would not rely on the library indicating an error.

Reply:
Good catch, I asked internal security library team about it, they confirmed this is an  issue and will fix it this week.

Reply:
As discussed, we can then expect the library to do the check once it is fixed. We should also add a comment in the documentation that ODP assumes that the library is compiled with SAFE_DATA or else crypto operations with too long cipher/auth ranges will have undefined behaviour.

With that: Reviewed-by: Janne Peltonen <janne.peltonen@nokia.com>